### DISCUSSION TITLE
Is OnDeactivate only called when explicit call Conductor.DeactivateItem?

### DISCUSSION THREAD

**Discussion by GF-Huang** [canton7/Stylet#258] [Q&A] [answered]
My `TabControl` is manual define items rather than `ItemsSource` binding. If I want my TabItem's view model OnDeactivate get called when the another `TabItem` selected, how can I do?

**Comment by canton7** [comment, marked as answer]
The `Conductor` classes all have special logic to activate and deactivate their children when appropriate. If you're not using this, you'll have to write that logic yourself.

One thing you might be able to do is:

```cs
public class ParentViewModel : Conductor.Collections.OneActive<IScreen>
{
    public AViewModel A { get; }
    public BViewModel B { get; }

    public ParentViewModel()
    {
        A = ...
        B = ...

        this.Items.AddRange(A, B);
    }
}
```
```xaml
<TabControl SelectedItem="{Binding ActiveItem}" .....
    <TabItem><ContentControl s:View.Model="{Binding A"}/></TabItem>
    <TabItem><ContentControl s:View.Model="{Binding B"}/></TabItem>
</TabControl>
```

(Untested).

So, your ViewModels exist as properties of the parent (which you bind to from your `TabControl`), but the parent is a conductor and those ViewModels are also added to the conductor's items. The `TabControl`'s `SelectedItem` is bound to the conductor's `ActiveItem`. When the current tab changes, the `TabControl` will set the `ActiveItem` on the conductor, which will activate/deactive its children as appropriate.

**Reply by GF-Huang** [reply]
Seems not work.

```
System.Windows.Data Error: 23 : Cannot convert 'System.Windows.Controls.TabItem 标头: 内容:' from type 'TabItem' to type 'Stylet.IScreen' for 'en-US' culture with default conversions; consider using Converter property of Binding. NotSupportedException:'System.NotSupportedException: ReferenceConverter cannot convert from System.Windows.Controls.TabItem.
   at System.ComponentModel.TypeConverter.GetConvertFromException(Object value)
   at System.ComponentModel.TypeConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
   at System.ComponentModel.ReferenceConverter.ConvertFrom(ITypeDescriptorContext context, CultureInfo culture, Object value)
   at MS.Internal.Data.DefaultValueConverter.ConvertHelper(Object o, Type destinationType, DependencyObject targetElement, CultureInfo culture, Boolean isForward)'
System.Windows.Data Error: 7 : ConvertBack cannot convert value 'System.Windows.Controls.TabItem 标头: 内容:' (type 'TabItem'). BindingExpression:Path=ActiveItem; DataItem='RootViewModel' (HashCode=55898385); target element is 'TabControl' (Name=''); target property is 'SelectedItem' (type 'Object') NotSupportedException:'System.NotSupportedException: ReferenceConverter cannot convert from System.Windows.Controls.TabItem.
   at MS.Internal.Data.DefaultValueConverter.ConvertHelper(Object o, Type destinationType, DependencyObject targetElement, CultureInfo culture, Boolean isForward)
   at MS.Internal.Data.ObjectTargetConverter.ConvertBack(Object o, Type type, Object parameter, CultureInfo culture)
   at System.Windows.Data.BindingExpression.ConvertBackHelper(IValueConverter converter, Object value, Type sourceType, Object parameter, CultureInfo culture)'
```

**Reply by canton7** [reply]
Ah yeah, the `TabControl` doesn't know about ViewModels at all, so the `SelectedItem` property will be one of the `TabItems`. 

Try `<TabControl SelectedValue="{Binding ActiveItem}" SelectedValuePath="DataContext" ...`

**Reply by GF-Huang** [reply]
![image](https://user-images.githubusercontent.com/4510984/118102679-ab821100-b40b-11eb-808d-2cf0cc5ffbbb.png)

**Reply by canton7** [reply]
Please put a little more effort into your messages, since you're asking for free help. I'm not sure what to take from the screenshot.

**Reply by GF-Huang** [reply]
Sorry friend, I had try <TabControl SelectedValue="{Binding ActiveItem}" SelectedValuePath="DataContext" , but not works. Looking for the runtime property inspector, I found the TabControl SelectedValue is bind to RootViewModel, it seems not works as your expected.

**Reply by canton7** [reply]
What do you mean exactly, "but not works"? What makes you think that it isn't working?

Are there any binding errors in the Output window?

What is the RootViewModel in your code? For all I know, RootViewModel could be the expected value of that property.

**Reply by GF-Huang** [reply]
No binding errors, "not works" means OnDeactivate not called.

```cs
public class RootViewModel : Conductor<IScreen>.Collection.OneActive { 
   ...
   public RootViewModel(AViewModel avm, BViewModel bvm) {
        Items.AddRange(new IScreen[] { avm, bvm });
   }
}
```

```xaml
<TabControl SelectedValue="{Binding ActiveItem}"
            SelectedValuePath="DataContext">
    <TabItem s:View.Model="{Binding AViewModel}" />
    <TabItem s:View.Model="{Binding BViewModel}" />
</TabControl>
```

**Reply by canton7** [reply]
Note, you don't have any code to set the `AViewModel` and `BViewModel` properties there.

Ah, the `s:View.Model` isn't setting the DataContext of your `TabItem`, just of its content. You might have to do `<TabItem DataContext="{Binding AViewModel}" s:View.Model="{Binding AViewModel}"/>`

**Reply by GF-Huang** [reply]
> Note, you don't have any code to set the AViewModel and BViewModel properties there.

I had set, but it is omitted here.

> Ah, the s:View.Model isn't setting the DataContext of your TabItem, just of its content. You might have to do <TabItem DataContext="{Binding AViewModel}" s:View.Model="{Binding AViewModel}"/>

The `DataContext="{Binding AViewModel}"` will breaking the `s:View.Model="{Binding AViewModel}"` because the `TabItem.DataContext` pointing to `AViewModel` rather than `RootViewModel`.

**Reply by GF-Huang** [reply]
This works, thank you!

`s:View.Model="{Binding DataContext.AViewModel, RelativeSource={RelativeSource AncestorType=TabControl}}"`
